util/topgen.py

```diff
@@ -724,7 +729,7 @@ def _process_top(topcfg, args, cfg_path, out_path, pass_idx):
     ips = search_ips(ip_dir)
 
     # exclude filtered IPs (to use top_${topname} one) and
-    exclude_list = generated_list + top_only_list
+    exclude_list = generated_list + [top_only_dict.keys()]
```

Comment:
This doesn't look quite right: you're appending an element which is a list of keys. Should it maybe be this?
```suggestion
    exclude_list = generated_list + list(top_only_dict.keys())
```

Reply:
o, sorry dumb python question, do the brackets not imply "list"?
because before i added the bracket, i did actually get a type error, but definitely using `list` is much more explicit.

Reply:
Brackets give a singleton list, but that's not what you want here. For example, if `top_only_dict` was `{'a': True, 'b': False}` and `generated_list` was `['x', 'y']` then your code would have given `['x', 'y', [some_object_representing(['a', 'b'])]` rather than `['x', 'y', 'a', 'b']`.

The `some_object_representing(...)` bit is because a `top_only_dict.keys()` doesn't actually construct a list: it returns an object that can be iterated over without needing to allocate and copy a new list. Much more efficient if the dictionary is enormous, but not quite the right type to add together.

Reply:
thanks for the explanation! i guess it just so happened i got some kind of iterable object, even though it was not a list. I've corrected this part.